src/nedap/speced/def/impl/spec_assertion.clj

```diff
@@ -20,15 +20,29 @@
        (catch ~(if clj?
                  `ExceptionInfo
                  'ExceptionInfo) e#
-         (let [spec# (:spec (ex-data e#))
-               inferred-specs# (set (map :spec (extract-specs-from-metadata {:tag ~spec-sym} ~clj?)))]
+         (let [{ex-data-spec#        :spec
+                ex-data-spec-object# :spec-object
+                ex-data-quoted-spec# :quoted-spec
+                :as                  ex-data-object#} (ex-data e#)
+               legacy?# (not (contains? ex-data-object# :spec-object))
+               spec# (if legacy?#
+                       ex-data-spec#
+                       ex-data-spec-object#)
+               inferred-specs# (->> (extract-specs-from-metadata {:tag ~spec-sym} ~clj?)
+                                    (map :spec)
+                                    set)]
 
            ;; rethrow if no spec failure is found
            (when-not spec#
              (throw e#))
 
-           (if (or (= ~spec-sym spec#)
-                   (contains? inferred-specs# spec#))
+           (if (or (if legacy?#
+                     (= ~spec-sym spec#)
+                     (or (= ex-data-quoted-spec# ~spec-sym)
+                         (= ex-data-spec-object# ~spec-sym)))
+                   (contains? inferred-specs# (if legacy?#
+                                                spec#
+                                                ex-data-quoted-spec#)))
```

Comment:
Here's my attempt at making this somewhat readible

```clj
(if (if legacy?#
       ((into #{~spec-sym} inferred-specs#) spec#)
       (or (#{ex-data-quoted-spec#, ex-data-quoted-object#} ~spec-sym)
           (contains? inferred-specs# ex-data-quoted-spec#))
```

Reply:
Nice one!

Reply:
I remembered now - building literal, multi-element sets/hashmaps in macros is dangerous. You can get `java.lang.IllegalArgumentException: Duplicate key ...`